### ISSUE TITLE
Create code style guide/settings

### ISSUE THREAD

**Issue by ctron** [eclipse/kura#324]
Maybe there already is, but I did not find anything in the GitHub Wiki, the Eclipse Wiki, the `CONTRIBUTING.md` file or the `README.md` file.

What are the requirements for code style when submitting changes? And were do I find those?

Maybe it would also be nice to have an Eclipse code formatter settings file, so that I can be applied automatically on each save.

**Comment by cdealti** [comment]
+1. Do you have suggestions?

**Comment by ctron** [comment]
In general it looks to me as if you are pretty close to the Eclipse standard profile. So I guess going with that would be fine.

Personally I don't really like it that much, but I guess you can argue about personal preferences on code style for ages :wink:

**Comment by cdealti** [comment]
We have style inconsistencies everywhere. Personally, I would just select all the code and hit Shift-Ctrl-F in Eclipse. However everyone has preferences on braces, tabs and spaces, line breaks, final keyword everywhere, s_static vs STATIC_ALL_UPPERCASE, m_.  I'd like to consider a formatter (and everyone should install Sonar Lint).

**Comment by ctron** [comment]
I am not sure if you can force everybody to install it.

I do know that everybody has preferences on code style :wink: But I would just say that the project decides on a code style. So that not everybody brings in his own style. Many projects do that. So it is a valid approach.

The way people do apply this style is a different topic. Manually? Fine! But the project could help in providing a settings file for Eclipse, IntelliJ and maybe other IDEs.

**Comment by amitjoy** [comment]
According to the Eclipse Code Style as provided in Eclipse IDE, I excerpted few important factors which would be great if followed.

- Add 'this' qualifier to unqualified field accesses
- Add 'this' qualifier to unqualified method accesses
- Convert control statement bodies to block
- Convert 'for' loops to enhanced 'for' loops
- Add final modifier to private fields
- Add final modifier to method parameters
- Add final modifier to local variables
- Use lambda where possible (_Not possible as of now_)
- Put expressions in parentheses
- Remove unused imports
- Remove unused local variables
- Add missing '@Override' annotations
- Add missing '@Override' annotations to implementations of interface methods
- Add missing '@Deprecated' annotations
- Remove unnecessary casts
- Remove unnecessary '$NON-NLS$' tags
- Sort all members
- Remove trailing white spaces on all lines
- Correct indentation
- Remove redundant type arguments
- Add file header

**Comment by ctron** [comment]
I wouldn't add a file adder automatically though. That sometimes goes wrong and may cause more issues than it helps.

**Comment by amitjoy** [comment]
I have been using **JAutodoc** in several projects since last year and it works fine. You can try and check it. It's really good.

**Comment by ctron** [comment]
I just pushed a change in order to provide a basis for discussion. I do think we need to move forward on this topic.

I would suggest that we first start with an Eclipse settings file for cleanup and code formatting and then formalize this in a document.

The code style is based on the Eclipse style with the following main differences:
- Use spaces instead of tabs
- The line width is 180 characters, of course you can still split lines before that
- Comment formatting is disabled
- Javadoc formatting is stripped down to a bare minimum, and lines are never joined

The code cleanup is limited to a set of less destructive rules:

- Add 'this' qualifier to unqualified field accesses
- Remove 'this' qualifier for non static method accesses
- Change non static accesses to static members using declaring type
- Change indirect accesses to static members to direct accesses (accesses through subtypes)
- Convert control statement bodies to block
- Add final modifier to private fields
- Add final modifier to method parameters
- Add final modifier to local variables
- Remove unnecessary parentheses
- Remove unused imports
- Add missing '@Override' annotations
- Add missing '@Override' annotations to implementations of interface methods
- Add missing '@Deprecated' annotations
- Add missing serial version ID (default 1L)
- Remove unnecessary casts
- Remove unnecessary '$NON-NLS$' tags
- Organize imports
- Format source code
- Remove trailing white spaces on all lines
- Correct indentation
- Remove redundant type arguments

---

I would suggest not to (of course this is all highly personal):
- Add 'this' qualifier to unqualified method accesses -> This really clutters up readability, and currently it is not done in Kura. And compared to field accesses (where you have local variables and parameters) you cannot run this kind of issue with methods.
- Use lambda where possible -> In general yes, but I think for the moment this is too much. Using lambdas is fine, but it should be the decision of the author.
- Sort all members -> This would completely screw up existing code.
- Add file header -> you can use the Eclipse "Releng" tool for doing that manually, it generates "nice" Eclipse style file headers.

**Comment by ctron** [comment]
I would also suggest to not use prefixed (m_, s_, ...) on variable names.

**Comment by ctron** [comment]
@mcarrer @dwoodard1 Please have a look at this and give some feedback so that we can move forward on this one.

**Comment by mcarrer** [comment]
Ok, we will review.

**Comment by MMaiero** [comment]
Proposed formatter and cleanup profiles agreed with Kapua team:
[Kura_ide_profiles.zip](https://github.com/eclipse/kura/files/469436/Kura_ide_profiles.zip)

**Comment by MMaiero** [comment]
After discussion during the weekly calls, we end up with a new version of the proposed Kura formatter: this consist in a set of modifications of the standard Eclipse formatter, in order to be simpler and adaptable for other IDEs. In [Kura Code Style.xlsx](https://github.com/eclipse/kura/files/487311/Kura.Kapua.Code.Style.xlsx) we addressed the differences between our proposed formatter and the Eclipse IDE one.

I've created a new branch named env_code-refactor that contains a reviewed [contributing.md](https://github.com/eclipse/kura/blob/dc5b6dd38e33cb602d8f75a43ef760ab472edb7f/CONTRIBUTING.md) file and where the proposed [cleanup](https://github.com/eclipse/kura/blob/16ef2e01e0b0f368fdb82c877f446798c4287232/kura/setups/formatting/KuraCleanupProfile.xml) and [formatter](https://github.com/eclipse/kura/blob/16ef2e01e0b0f368fdb82c877f446798c4287232/kura/setups/formatting/KuraFormatter.xml) profiles are available.

If no issues or changes will be proposed by the community, we plan to entirely format the Kura code on Friday 09/30/2016.
From that point on, all the pull requests **must** comply the accepted cleanup and formatter profiles.

**Comment by amitjoy** [comment]
I believe the Code Style for Kura should not obligate developers to always use `Add 'this' qualifier to unqualified field accesses`. I personally prefer adding `this` to unqualified field accesses as it makes it very easy for new developers to distinguish between local variable and field access. On the contrary, it makes code a bit verbose in case of inner classes and local classes.

```
class A {
        class B {
            public void doIt() {
                A.this.a = 5;
            }
        }
        int a = 10;
    }
```
It should be left to the developer of the code to use `this` for unqualified field access. It's just my idea. Instead we can optionally use it in case it is really needed or just leave it to the developer.

**Comment by MMaiero** [comment]
I agree with @amitjoy: trying the refactoring and cleanup profiles with the new Cloud tab branch, the code becomes really nasty and difficult to read. An example:
`private void getCloudServicePidRegex(final String factoryPid) {
        this.gwtCloudService.getCloudServicePidRegex(factoryPid, new AsyncCallback<String>() {

            @Override
            public void onFailure(Throwable caught) {
                FailureHandler.handle(caught, CloudInstancesUi.this.gwtCloudService.getClass().getSimpleName());
            }

            @Override
            public void onSuccess(String result) {
                if (result != null) {
                    CloudInstancesUi.this.regexValidator = new RegExValidator(result);
                } else {
                    CloudInstancesUi.this.regexValidator = new RegExValidator(".+");
                }
                CloudInstancesUi.this.cloudServicePid.addValidator(CloudInstancesUi.this.regexValidator);
                CloudInstancesUi.this.cloudServicePid.addKeyUpHandler(new KeyUpHandler() {

                    @Override
                    public void onKeyUp(KeyUpEvent event) {
                        CloudInstancesUi.this.cloudServicePid.validate();
                    }
                });
                CloudInstancesUi.this.cloudServicePid.addBlurHandler(new BlurHandler() {

                    @Override
                    public void onBlur(BlurEvent event) {
                        CloudInstancesUi.this.cloudServicePid.validate();
                    }
                });
            }`

I would prefer to leave the developer the possibility to set the this, if needed instead of forcing it.
@ctron what you think?

**Comment by ctron** [comment]
The problem is see is, that this really prevents people from introduction bugs. It so easy to miss that. If people think about that, it is less likely to happen. If they don't, they tend to make exactly those kinds of mistakes.

So letting the developer choose on this one actually defies its purpose, catching bugs which people do not think about. 

I agree, it makes the code more verbose. But looking at the Web UI code, I feel the urge to split that up anyway. And making up that decision just due the API of GWT which is only used in one bundle, seems wrong to me.

**Comment by Coduz** [comment]
Hi @ctron,

you are talking about bugs that can be introduced by not using `this`, right?
I'm not an experienced developer as you probably are, but it never happened to me to encounter bugs or problems by not adding `this`. Moreover I'm not able to see or imagine which kind of mistakes you can fall in by not adding it. 

As far as i can see it is improbable to easily encounter those problems and I would like to have a readable code instead,

Regards,

_Alberto_

**Comment by ctron** [comment]
@Coduz So maybe I have seen more code than you :wink: 

In general a field may be hidden by a parameter or local variable. You can imaging all kinds of bugs which that may cause.

I will give you some examples which pop into my mind:

```java
public void setFoo ( int foo) {
  foo = foo;
 // should be this.foo = foo
 // would also be caught by making parameter foo final
}

public void updateFoo ( Integer foo )  {
   if ( foo == null ) {
      foo = 1; // should be this.foo
   } else  {
      this.foo = foo; 
  }
}

// only set service is service is not already set
public void makeFoo () {
   Service service = new Service ();
   if ( service == null ) { // ops, should be this.service
      this.service = service;   
   }
}
```

So if you think a bit about this, you will come up with all kind of other scenarios where stuff like this could happen.

**Comment by Coduz** [comment]
Hi @ctron ,

uhm... I see, but this is a situation in which any IDE warns you that you are hiding properties (unless you're using nano or vim 😆 😆 😆 ) and you can rapidly fix it.

I think that this problem is overrated. The chance of problems that will require a lot of time to resolve is negligible. I would like to keep it as a developer preference or force it only when is necessary. In this way we will have a less verbose code and this is good, as you said, without any major downside.

Regards, 

_Alberto_

**Comment by ctron** [comment]
How does your IDE warn you about:

```java
public void updateFoo ( Integer foo )  {
   if ( foo == null ) {
      foo = 1; // should be this.foo
   } else  {
      this.foo = foo; 
  }
}
```

And of course, this is easily fixed, if you are aware of it, which is the issue in the first place. But if you are unaware of it, mistakes happen.

I think that is better to accept a little bit less "stylish" code, or a little bit more verbose code than to accept possible bugs.

The point is, if you enforce this, and just always require it, then it cannot happen that bugs come along over time. Just requesting this "if necessary" will exactly bring bugs like this.

**Comment by Coduz** [comment]
Yes you are right on that one, there is no warning, but (using Eclipse at least) I can clearly see the difference between blue and black writings :nerd_face: 
It can be the case that not everyone sees it, anyway :) 

I was exposing my opinion on the fact that it is very difficult that all this bugs can occurs :) 

Regards,

_Alberto_

**Comment by amitjoy** [comment]
@ctron To avoid such situation, the developer can use `final` for the parameters and I personally like using `final` for parameters and local variables. But my personal opinion is to leave it for developers rather forcing it.

**Comment by ctron** [comment]
@amitjoy I agree :smile:  I do like `final` as well. I I would also enforce the use of that, as you know :wink: But there we had the same discussion.

And in the end, it is not an issue if the developer thinks about such cases and decides to use against it. Problems arise when developers to not think about it at all and wonder why it is not working.

So all that I am saying is, that from a project perspective, just enforcing this will keep bugs away. Where as if people may choose whether to add `final` and `this` as they like, you will end up with a mixture of both throughout the code and search some bugs which could be avoided from the beginning.

**Comment by ctron** [comment]
Just as an example of what I mean, have a look at `org.eclipse.kura.core.deployment.download.impl.HttpDownloadCountingOutputStream`

Pay attention to `is`:

```java
urlConnection.setConnectTimeout(connectTimeout);
urlConnection.setReadTimeout(readTimeout);

testConnectionProtocol(urlConnection);

is = localUrl.openStream();

String s = urlConnection.getHeaderField("Content-Length");
s_logger.info("Content-lenght: " + s);
```

And now have a look at the source code, and please tell me a reason why "is" is a field. If, by saving the file, it would turn into:

```java
HttpDownloadCountingOutputStream.this.is = localUrl.openStream();
```

Someone would scratch his head and wonder why. So far, no IDE has warned about that one.

**Comment by MMaiero** [comment]
I've proceeded to the formatting and cleanup operations. 
During this step I've had to slightly modify the formatter to increase the readability of enums. In particular, in 'enum' declaration -> constants, I've set:

- Line wrapping to all elements, every element on a new line

- Force split

- Default indentation

Due to these changes I've submitted the modifications as a PR #584 waiting for feedback and approval to perform the merge.
